File: API/2327_WlxBW5Zone/CS/WlxBw5ZoneStrategy.cs
using System;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Bill Williams zone strategy based on Awesome and Accelerator oscillators.
/// Opens or reverses positions after five consecutive oscillator moves.
/// </summary>
public class WlxBw5ZoneStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<bool> _direct;
	private readonly StrategyParam<int> _signalBar;

	private AwesomeOscillator _ao;
	private AcceleratorOscillator _ac;

	private decimal? _ao0, _ao1, _ao2, _ao3, _ao4;
	private decimal? _ac0, _ac1, _ac2, _ac3, _ac4;
	private bool _flagUp;
	private bool _flagDown;

	/// <summary>
	/// Candle type used by the strategy.
	/// </summary>
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	/// <summary>
	/// Direction flag. If false signals are reversed.
	/// </summary>
	public bool Direct { get => _direct.Value; set => _direct.Value = value; }

	/// <summary>
	/// Signal bar shift.
	/// </summary>
	public int SignalBar { get => _signalBar.Value; set => _signalBar.Value = value; }

	/// <summary>
	/// Initializes a new instance of <see cref="WlxBw5ZoneStrategy"/>.
	/// </summary>
	public WlxBw5ZoneStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(4).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");

		_direct = Param(nameof(Direct), true)
			.SetDisplay("Direct", "Use direct signals", "General");

		_signalBar = Param(nameof(SignalBar), 1)
			.SetRange(0, 5)
			.SetDisplay("Signal Bar", "Bar shift for signals", "General");
	}

	/// <inheritdoc />
	public override System.Collections.Generic.IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_ao0 = _ao1 = _ao2 = _ao3 = _ao4 = null;
		_ac0 = _ac1 = _ac2 = _ac3 = _ac4 = null;
		_flagUp = false;
		_flagDown = false;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		_ao = new AwesomeOscillator();
		_ac = new AcceleratorOscillator();

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(_ao, _ac, ProcessCandle).Start();
	}

	private void ProcessCandle(ICandleMessage candle, decimal ao, decimal ac)
	{
		if (candle.State != CandleStates.Finished)
			return;

		// shift previous values
		_ao4 = _ao3; _ao3 = _ao2; _ao2 = _ao1; _ao1 = _ao0; _ao0 = ao;
		_ac4 = _ac3; _ac3 = _ac2; _ac2 = _ac1; _ac1 = _ac0; _ac0 = ac;

		if (_ao4 is null || _ac4 is null)
			return; // not enough data

		var isUpSeq = _ao0 > _ao1 && _ao1 > _ao2 && _ao2 > _ao3 && _ao3 > _ao4 &&
			_ac0 > _ac1 && _ac1 > _ac2 && _ac2 > _ac3 && _ac3 > _ac4;

		var isDownSeq = _ao0 < _ao1 && _ao1 < _ao2 && _ao2 < _ao3 && _ao3 < _ao4 &&
			_ac0 < _ac1 && _ac1 < _ac2 && _ac2 < _ac3 && _ac3 < _ac4;

		if (!_flagUp && isUpSeq)
		{
			if (Direct)
			{
				if (Position <= 0)
					BuyMarket();
			}
			else
			{
				if (Position >= 0)
					SellMarket();
			}

			_flagUp = true;
		}

		if (!_flagDown && isDownSeq)
		{
			if (Direct)
			{
				if (Position >= 0)
					SellMarket();
			}
			else
			{
				if (Position <= 0)
					BuyMarket();
			}

			_flagDown = true;
		}

		if (_ao0 <= _ao1 || _ac0 <= _ac1)
			_flagUp = false;

		if (_ao0 >= _ao1 || _ac0 >= _ac1)
			_flagDown = false;
	}
}

File: API/2301-2400/2327_WlxBW5Zone/PY/wlx_bw5_zone_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import AwesomeOscillator, SimpleMovingAverage
from StockSharp.Algo.Strategies import Strategy
from indicator_extensions import *

class wlx_bw5_zone_strategy(Strategy):
    def __init__(self):
        super(wlx_bw5_zone_strategy, self).__init__()
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._direct = self.Param("Direct", True) \
            .SetDisplay("Direct", "Use direct signals", "General")
        self._signal_bar = self.Param("SignalBar", 1) \
            .SetDisplay("Signal Bar", "Bar shift for signals", "General")
        self._ao = None
        self._ao_sma = None
        self._ao0 = None
        self._ao1 = None
        self._ao2 = None
        self._ao3 = None
        self._ao4 = None
        self._ac0 = None
        self._ac1 = None
        self._ac2 = None
        self._ac3 = None
        self._ac4 = None
        self._flag_up = False
        self._flag_down = False

    @property
    def candle_type(self):
        return self._candle_type.Value

    @property
    def direct(self):
        return self._direct.Value

    @property
    def signal_bar(self):
        return self._signal_bar.Value

    def OnReseted(self):
        super(wlx_bw5_zone_strategy, self).OnReseted()
        self._ao = None
        self._ao_sma = None
        self._ao0 = self._ao1 = self._ao2 = self._ao3 = self._ao4 = None
        self._ac0 = self._ac1 = self._ac2 = self._ac3 = self._ac4 = None
        self._flag_up = False
        self._flag_down = False

    def OnStarted2(self, time):
        super(wlx_bw5_zone_strategy, self).OnStarted2(time)
        self._ao0 = self._ao1 = self._ao2 = self._ao3 = self._ao4 = None
        self._ac0 = self._ac1 = self._ac2 = self._ac3 = self._ac4 = None
        self._flag_up = False
        self._flag_down = False
        self._ao = AwesomeOscillator()
        self._ao_sma = SimpleMovingAverage()
        self._ao_sma.Length = 5
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(self._ao, self.process_candle).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, self._ao)
            self.DrawOwnTrades(area)

    def process_candle(self, candle, ao):
        if candle.State != CandleStates.Finished:
            return
        ao = float(ao)
        sma_result = process_float(self._ao_sma, ao, candle.CloseTime, True)
        ac = ao - float(sma_result)
        self._ao4 = self._ao3
        self._ao3 = self._ao2
        self._ao2 = self._ao1
        self._ao1 = self._ao0
        self._ao0 = ao
        self._ac4 = self._ac3
        self._ac3 = self._ac2
        self._ac2 = self._ac1
        self._ac1 = self._ac0
        self._ac0 = ac
        if self._ao4 is None or self._ac4 is None:
            return
        is_up_seq = (self._ao0 > self._ao1 and self._ao1 > self._ao2 and
                     self._ao2 > self._ao3 and self._ao3 > self._ao4 and
                     self._ac0 > self._ac1 and self._ac1 > self._ac2 and
                     self._ac2 > self._ac3 and self._ac3 > self._ac4)
        is_down_seq = (self._ao0 < self._ao1 and self._ao1 < self._ao2 and
                       self._ao2 < self._ao3 and self._ao3 < self._ao4 and
                       self._ac0 < self._ac1 and self._ac1 < self._ac2 and
                       self._ac2 < self._ac3 and self._ac3 < self._ac4)
        if not self._flag_up and is_up_seq:
            if self.direct:
                if self.Position <= 0:
                    self.BuyMarket()
            else:
                if self.Position >= 0:
                    self.SellMarket()
            self._flag_up = True
        if not self._flag_down and is_down_seq:
            if self.direct:
                if self.Position >= 0:
                    self.SellMarket()
            else:
                if self.Position <= 0:
                    self.BuyMarket()
            self._flag_down = True
        if self._ao0 <= self._ao1 or self._ac0 <= self._ac1:
            self._flag_up = False
        if self._ao0 >= self._ao1 or self._ac0 >= self._ac1:
            self._flag_down = False

    def CreateClone(self):
        return wlx_bw5_zone_strategy()
